lib/Doctrine/ORM/LazyCriteriaCollection.php

```diff
@@ -116,9 +116,30 @@ public function contains($element)
      */
     public function matching(Criteria $criteria)
     {
-        $this->initialize();
+        if ($this->isInitialized()) {
+            return $this->collection->matching($criteria);
+        }
+
+        $mergedCriteria = clone $this->criteria;
+
+        if (null !== $criteria->getWhereExpression()) {
+            $mergedCriteria->andWhere($criteria->getWhereExpression());
+        }
+
+        if (null !== $criteria->getFirstResult()) {
+            $mergedCriteria->setFirstResult($criteria->getFirstResult());
+        }
+
+        if (null !== $criteria->getMaxResults()) {
+            $mergedCriteria->setMaxResults($criteria->getMaxResults());
```

Comment:
Maybe constraints should be nested?

`
    $maxResults = $mergedCriteria->getMaxResults() + $criteria->getMaxResults();
    $mergedCriteria->setMaxResults($maxResults);
`